### ISSUE TITLE
`az login --use-device-code` fails: AADSTS70020: The provided value for the input parameter 'device_code' is not valid. This device code has expired.

### ISSUE THREAD

**Issue by shaygover** [Azure/azure-cli#22251]
## Describe the bug

**Command Name**
`az login`

**Errors:**
```
AADSTS70020: The provided value for the input parameter 'device_code' is not valid. This device code has expired.
Trace ID: a8f3578e-d4cb-44ac-91a8-64c1322d6f00
Correlation ID: ed438693-0685-4cfe-923e-b73b5cd759e5
Timestamp: 2022-05-03 08:56:46Z

```

## To Reproduce:
Steps to reproduce the behavior. Note that argument values have been redacted, as they may contain sensitive information.

- _Put any pre-requisite steps here..._
- `az login -t {} --use-device-code`

Tried many times. Worked fine last week so I tried with 2.35.0. Same result.

## Expected Behavior
Succesfull login. Or at least a device code that doesn't expire after 1-5 seconds.

## Environment Summary
```
Linux-5.17.5-arch1-1-x86_64-with-glibc2.35, Arch Linux
Python 3.10.4
Installer: 

azure-cli 2.36.0 and 2.35.0
```

**Comment by yonzhan** [comment]
@jiasli for awareness

**Comment by tschechniker** [comment]
i'm having the same issue: https://github.com/Azure/azure-cli/issues/22252 seems like i wrote while you created the issue 👍 

This is critical for me and a current production showstopper.

**Comment by jiasli** [comment]
I followed https://docs.microsoft.com/en-us/azure/active-directory/develop/reference-aadsts-error-codes and visited https://login.microsoftonline.com/error to check `AADSTS70020`, but got no further info:

Error Code | 70020
-- | --
Message | The provided value for the input parameter 'device_code' is not valid. This device code has expired.
Remediation | Please make a new Device Authorization Request.

I debugged into MSAL's code

https://github.com/AzureAD/microsoft-authentication-library-for-python/blob/14965f76b02bd8544cf2f68162e01806846e1bbf/msal/oauth2cli/oauth2.py#L330

and saw [`expires_in` defined in device code flow](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-device-code#device-authorization-response) is by default 900s.

![image](https://user-images.githubusercontent.com/4003950/166436194-ad9b22b2-7449-47e9-9433-00ec1a438a18.png)

That should give the user enough time to login. I can't reproduce with `organizations` tenant, which for my account is `72f988bf-86f1-41af-91ab-2d7cd011db47`.

**Comment by tschechniker** [comment]
So how can we proceed here? I can reproduce it on all my subscriptions and tenants. Is there any region specific error? For myself i talk to westeurope. @shaygover whats your region?

**Comment by jiasli** [comment]
As https://github.com/Azure/azure-cli/issues/22252#issue-1223890576 suggested, this issue also happens for Azure PowerShell cmdlet `Connect-AzAccount -UseDeviceAuthentication`, so I doubt this is a service issue.

Could you directly get a device code and share the result?

```
curl -d 'client_id=04b07795-8ddb-461a-bbee-02f9e1bf7b46&scope=https%3A%2F%2Fmanagement.core.windows.net%2F%2F.default+offline_access+openid+profile' https://login.microsoftonline.com/organizations/oauth2/v2.0/devicecode

Example response:
{"user_code":"xxx","device_code":"xxx","verification_uri":"https://microsoft.com/devicelogin","expires_in":900,"interval":5,"message":"To sign in, use a web browser to open the page https://microsoft.com/devicelogin and enter the code xxx to authenticate."}
```

See https://curl.se/docs/manual.html#post-http

**Comment by tschechniker** [comment]
Response (user and device code has been deleted but seemd to had valid values):

{"user_code":"XXX","device_code":"XXX","verification_uri":"https://microsoft.com/devicelogin","expires_in":900,"interval":5,"message":"To sign in, use a web browser to open the page https://microsoft.com/devicelogin and enter the code XXX to authenticate."}

**Comment by shaygover** [comment]
@jiasli With curl I get:
`"user_code":"D3YUBE53L","device_code":"DAQABAAEAAAD--DLA3VO7QrddgJg7WevrL1HwUa0pNrxhFgF5z_2R2Ohb-v2x18QRuiUOX8cRGG2T_8MJKHWoK6P4PAhVMW_MSH81Wj7nxX1KVFuT0RN-am9rvuMfr6K-do_rI4SR72SXAOCXVSQfpCYEri6jNBx1xWJ0dEe2xPaVnPGVMLD68enHuGnxv7robPCf0P-Ch7UgAA","verification_uri":"https://microsoft.com/devicelogin","expires_in":900,"interval":5,"message":"To sign in, use a web browser to open the page https://microsoft.com/devicelogin and enter the code D3YUBE53L to authenticate."}`

@tschechniker West Europe

**Comment by jiasli** [comment]
I saw an incident has already been opened for this issue. I will check with AAD team and let you know if there is any update.

Meanwhile, could you also run the command with `-i` which will show more info about the server?

```
$ curl -d 'client_id=04b07795-8ddb-461a-bbee-02f9e1bf7b46&scope=https%3A%2F%2Fmanagement.core.windows.net%2F%2F.default+offline_access+openid+profile' https://login.microsoftonline.com/organizations/oauth2/v2.0/devicecode -i
HTTP/1.1 200 OK
Cache-Control: no-store, no-cache
Pragma: no-cache
Content-Type: application/json; charset=utf-8
Expires: -1
Strict-Transport-Security: max-age=31536000; includeSubDomains
X-Content-Type-Options: nosniff
P3P: CP="DSP CUR OTPi IND OTRi ONL FIN"
x-ms-request-id: e44170d4-f198-4612-bffd-864786490100
x-ms-ests-server: 2.1.12651.9 - KRSLR1 ProdSlices
X-XSS-Protection: 0
Set-Cookie: fpc=AjQrn3OX4ZpMrEh-DcSa_KN1eSrYAQAAAI4CA9oOAAAA; expires=Thu, 02-Jun-2022 11:01:34 GMT; path=/; secure; HttpOnly; SameSite=None
Set-Cookie: esctx=AQABAAAAAAD--DLA3VO7QrddgJg7Wevrbw5Gj3oEcz48n1UFHRDjhlPUW-VM6cAuMkNDQHDdJ2YFkDzcsbGEed3xwdN4LKO06EBt22_LPpLUVKEaB-1nBIU2D2uTZL-qSKzHMgVz0WI2vqjiORS7wR4yc-e-r69jljQArnK3EMI7bExXist-PhAmtYZ4p7E4yHX6O5g_9McgAA; domain=.login.microsoftonline.com; path=/; secure; HttpOnly; SameSite=None
Set-Cookie: x-ms-gateway-slice=estsfd; path=/; secure; samesite=none; httponly
Set-Cookie: stsservicecookie=estsfd; path=/; secure; samesite=none; httponly
Date: Tue, 03 May 2022 11:01:34 GMT
Content-Length: 473

{"user_code":"xxx","device_code":"xxx","verification_uri":"https://microsoft.com/devicelogin","expires_in":900,"interval":5,"message":"To sign in, use a web browser to open the page https://microsoft.com/devicelogin and enter the code xxx to authenticate."}
```

**Comment by shaygover** [comment]
@jiasli 
`HTTP/1.1 200 OK
Cache-Control: no-store, no-cache
Pragma: no-cache
Content-Type: application/json; charset=utf-8
Expires: -1
Strict-Transport-Security: max-age=31536000; includeSubDomains
X-Content-Type-Options: nosniff
P3P: CP="DSP CUR OTPi IND OTRi ONL FIN"
x-ms-request-id: c7a5c72a-ecd5-40b2-b8c1-5914013e0600
x-ms-ests-server: 2.1.12651.9 - WEULR1 ProdSlices
X-XSS-Protection: 0
Set-Cookie: fpc=Am8XuADE3lFHjMJ3EGYfSg91eSrYAQAAAGMDA9oOAAAA; expires=Thu, 02-Jun-2022 11:05:07 GMT; path=/; secure; HttpOnly; SameSite=None
Set-Cookie: esctx=AQABAAAAAAD--DLA3VO7QrddgJg7WevrIG9QnEQyyDQZb_b12pHYgyTXXQRAYG2CHsywek8zOk00CdzULoYvDZFn0Uy91uUdAe1-US1X2WwkEyFGd-NhaDcA5M1p0J8Wu7ZzMlAyK-DGyoX6Wj7YYLbu3UrViktRg1bQsWMIRJ24NbmZAeaJAwPTbrcziJwjVoGiTqz5dxsgAA; domain=.login.microsoftonline.com; path=/; secure; HttpOnly; SameSite=None
Set-Cookie: x-ms-gateway-slice=estsfd; path=/; secure; samesite=none; httponly
Set-Cookie: stsservicecookie=estsfd; path=/; secure; samesite=none; httponly
Date: Tue, 03 May 2022 11:05:06 GMT
Content-Length: 473

{"user_code":"CDDGY2ACF","device_code":"CAQABAAEAAAD--DLA3VO7QrddgJg7Wevrpah6eQPkUGvGZwzG8O67c--91wjj6geO9c_b3vRKW0uOHcNPrYcq3IAwpQp6HCmEPbWvRDDoCzW2u1o5UsI8EgOG3SNCECvY0EVZ_JSeKFQuBlPLFEDsppGUG2jxwewXJTXJSsWpW86vRfP89Q1_IJ4VS7xVylA-2_N6Pk30EaMgAA","verification_uri":"https://microsoft.com/devicelogin","expires_in":900,"interval":5,"message":"To sign in, use a web browser to open the page https://microsoft.com/devicelogin and enter the code CDDGY2ACF to authenticate."}`

**Comment by tschechniker** [comment]
HTTP/1.1 200 OK
Cache-Control: no-store, no-cache
Pragma: no-cache
Content-Type: application/json; charset=utf-8
Expires: -1
Strict-Transport-Security: max-age=31536000; includeSubDomains
X-Content-Type-Options: nosniff
P3P: CP="DSP CUR OTPi IND OTRi ONL FIN"
x-ms-request-id: a3237540-9a73-4bcc-95c4-13ccf0e74600
x-ms-ests-server: 2.1.12651.9 - WEULR2 ProdSlices
X-XSS-Protection: 0
Set-Cookie: fpc=Aqr1evQXVSdFk6gOTNq3opJ1eSrYAQAAAMkDA9oOAAAA; expires=Thu, 02-Jun-2022 11:06:50 GMT; path=/; secure; HttpOnly; SameSite=None
Set-Cookie: esctx=AQABAAAAAAD--DLA3VO7QrddgJg7WevrqE4_0l8E9_B1GIOcxLJBnUru0t5xMYc1sHYM8MsxQ4SaKtESQ_a8MOQxJxhDvGc8Pnmf6zGCW0S9Fa9OV8j6PROMTmhmvT4YD7nSJKsTXQxLUR4orxuQzOUxgWKR3041D9NhwxiMkcF274qQ1-ZtxvLcF6hVM9FP2YEX5IdiapUgAA; domain=.login.microsoftonline.com; path=/; secure; HttpOnly; SameSite=None
Set-Cookie: x-ms-gateway-slice=estsfd; path=/; secure; samesite=none; httponly
Set-Cookie: stsservicecookie=estsfd; path=/; secure; samesite=none; httponly
Date: Tue, 03 May 2022 11:06:49 GMT
Content-Length: 473

{"user_code":"XXX","device_code":"XXX","verification_uri":"https://microsoft.com/devicelogin","expires_in":900,"interval":5,"message":"To sign in, use a web browser to open the page https://microsoft.com/devicelogin and enter the code XXX to authenticate."}

**Comment by jiasli** [comment]
Judging from

```
x-ms-ests-server: 2.1.12651.9 - WEULR2 ProdSlices
```

This should be happening for westeurope. I will keep you updated as soon as there is any updates.

**Comment by yonzhan** [comment]
This issue affected all client tools rather than Azure CLI and needs to be fixed by AAD service.

**Comment by usrme** [comment]
Is there a link to another issue where we can follow the status?

**Comment by jiasli** [comment]
We are working internally with ESTS team with highest priority. We will keep you updated in this GitHub issue.

**Comment by mbsnl** [comment]
As of 14:09 UTC, it seems to be solved for me (Europe).
I reported an official MS Case against the AzureAD Service, they confirmed this being an AzureAD issue, but I do not know if they declared it solved or not.

**Comment by jiasli** [comment]
This issue has been mitigated by AAD service team. Please retry `az login --use-device-code` and let us know if you can get unblocked.

**Comment by usrme** [comment]
No longer having issues! :heavy_check_mark:

**Comment by ai2create** [comment]
> This issue has been mitigated by AAD service team. Please retry `az login --use-device-code` and let us know if you can get unblocked.

This works thanks
